proposals/2024-09-25_metadata-labels.md

```diff
@@ -0,0 +1,186 @@
+
+Users don't see the `__type__` or `__unit__` labels in the Prometheus UI next to other labels by default.
+
+Users see no change to exposition formats as a result of this proposal.
+
+## How
+
+### PromQL Changes
+
+When a query for a metric returns multiple metrics with a different `__type__` or `__unit__` label, but the same `__name__`, an info annotation will be returned with the PromQL response, which is otherwise unmodified.
+
+Aggregations and label matches ignore `__unit__` and `__type__` and any operation removes the `__unit__` and `__type__` label (with the exception of `label_replace`).
+
+### Prometheus UI Changes
+
+When displaying a metric's labels in the table or in the graph views, the UI will hide labels starting with `__` (double underscore) by default, similar to the current handling of `__name__`. A "Show System Labels" check-box will be added, which shows hidden labels when checked.
+
+### Prometheus Server Ingestion
+
+When receiving OTLP or PRW 2.0, or when scraping the text, OM, or proto formats, the type and unit of the metric are added as the `__type__` and `__unit__` labels, if not already present.
+
+PRW 1.0 is omitted because metadata is sent separately from timeseries, making it infeasible to add the labels at ingestion time.
+
+Users can modify the type and unit of a metric at ingestion time by using `metric_relabel_configs`, and relabeling the `__type__` and `__unit__` labels.
+
+### Implementation Plan
+
+#### Milestone 1: Feature flag for adding labels
+
+Add a feature flag: `--enable-feature=identifying-type-and-unit`. When enabled, `__type__` and `__unit__` labels are added when receiving OTLP or PRW 2.0, or when scraping the text, OM, or proto formats. Implement any changes required to allow relabeling `__type__` and `__unit__` in `metric_relabel_configs`.
+
+#### Milestone 2: UI and PromQL changes
+
+During this stage, implement the UI and PromQL changes above. Iterate based on feedback. Changes should have no effect unless the identifying-type-and-unit feature flag is enabled.
+
+#### Milestone 3: Add NoNameChanges option for OTLP translation
+
+Add an option, `NoNameChanges` for the OTLP translation strategy. When enabled, it disables UTF-8 sanitization and the addition of suffixes. In the documentation for this option, recommend that `--enable-feature=identifying-type-and-unit` is enabled.
+
+## Alternatives
+
+### “Real” Type and Unit suffixes in **name**
+
+Similar to suffixes of _<unit> and _<type/total> but make it an explicit suffix using a delimiter not currently permitted in metric names. Specifying suffixes is optional when querying for a metric. When the type or unit suffix is omitted from a query, it would (design TBD) return results which include any type or unit suffix which exists for that name.
+
+NOTES:
+
+* `~` for units and `.` for type is just one example, there might be better operators/characters to use.
```

Comment:
If we have a fixed set of allowed strings for the `__type__`, we could probably declare them "reserved" and not usable for units. Then we could use the same special separator characters for both type and unit (`http_request_duration.seconds.histogram`, equivalent to `http_request_durations.histogram.seconds`). (Pitfall: If we extend the list of allowed types later, we might hit strings that have been used as units somewhere.)

Reply:
Could we move this part to another proposal? I agree that this is a separate thing from the original proposal here, and getting consensus would delay the work on the label's part 😬

Reply:
On the contrary. An objection against this proposal was that the label syntax is so complicated that people will stop including any hints about the type and unit when they write PromQL. At least mentioning the option of (maybe later) add an easier syntax for type and unit will _expedite_, rather than delay the acceptance of this proposal.

Reply:
Mentioned so we should be good to unblock this and propose extended PromQL syntax in a separate proposal. Thanks!